File: Java/graphs/DirectedCycleDetection.java
/**
Cycle detection in a directed graph
using dfs. Here, ststus array shows if 
the node is in the path that we are 
currently on. This is used to detect 
any presence of back edges.
backegde present <=> cycle present 
**/

import java.io.*; 
import java.util.*;

public class DirectedCycleDetection {
	private static boolean cycle;
	private static void dfs_visit(ArrayList<ArrayList<Integer> > graph, int src, boolean[] status, boolean[] visit) {
		visit[src] = true;
		status[src] = true;
		for(int i = 0; i < graph.get(src).size(); i++)
			if(visit[graph.get(src).get(i)] == false)
			    dfs_visit(graph, graph.get(src).get(i), status, visit);
			else
			    if(status[graph.get(src).get(i)] == true)
				    cycle = true;
	    status[src] = false;
	}

	private static void dfs(ArrayList<ArrayList<Integer> > graph, int v, boolean[] status) {
		boolean[] visit = new boolean[v];
		for(int i = 0; i < v; i++)
			if(visit[i] == false)
				dfs_visit(graph, i, status, visit);
	}

	private static void detect_cycle(ArrayList<ArrayList<Integer> > graph, int v) {
		boolean[] status = new boolean[v];
		dfs(graph, v, status);
		if(cycle == false)
		    System.out.println("No cycle exits in the given graph");
	    else
		    System.out.println("Cycle exists in the given graph");
	}

	public static void main(String[] args) {
		cycle = false;
		int v, e, a, b;
		// all vertices are labelled from 0 to v-1
		Scanner sc= new Scanner(System.in);
		v = sc.nextInt();
		e = sc.nextInt();
		ArrayList<ArrayList<Integer> > graph = new ArrayList<ArrayList<Integer> >(v);
		for (int i = 0; i < v; i++) 
            graph.add(new ArrayList<Integer>());
		// all directed edges
		for(int i = 0; i < e; i++) {
			a = sc.nextInt();
			b = sc.nextInt();
			// edge a -> b
			graph.get(a).add(b);
		} 
		detect_cycle(graph, v);
	}
}

/**
Input :
6 8
0 3
0 4
5 0
1 5
1 0
2 1
3 4
4 5
Output :
Cycle exists in the given graph

Time Complexity : O(v+e)
Space Complexity : O(v)
**/

File: C-Plus-Plus/graphs/DirectedCycleDetection.cpp
/**
Cycle detection in a directed graph
using DFS. Here, status array shows if the node is in the path that we are currently on. This is used to detect any presence of back edges.
backedge present <=> cycle present
**/

#include <bits/stdc++.h>
using namespace std;

bool cycle;

void dfs_visit(vector<vector<int>>& Graph, int src, vector<bool>& status, vector<bool>& visit) {
	// mark the node as visited
	visit[src] = 1;
	// mark the node saying that it is in the current path
	status[src] = 1;
	for (int i = 0; i < Graph[src].size(); i++)
		if (visit[Graph[src][i]] == 0)
			dfs_visit(Graph, Graph[src][i], status, visit);
		else if (status[Graph[src][i]] == 1)
			cycle = 1;
	// unmark the node saying the path which contains the node is done
	status[src] = 0;
}

void dfs(vector<vector<int>>& Graph, int vertex, vector<bool>& status) {
	// initially mark all nodes as not visited
	vector<bool> visit(vertex);
	// go on visiting each not visited node
	for (int i = 0; i < vertex; i++)
		if (visit[i] == 0)
			dfs_visit(Graph, i, status, visit);
}

void detect_cycle(vector<vector<int>>& Graph, int vertex) {
	vector<bool> status(vertex);
	dfs(Graph, vertex, status);
	if (cycle == 0)
		cout << "No cycle exits in the given graph \n";
	else
		cout << "Cycle exists in the given graph \n";
}

int main() {
	cycle = 0;
	int vertex, edge, a, b;
	// all vertices are labelled from 0 to v-1
	cin >> vertex >> edge;
	vector<vector<int>> Graph(vertex);
	// all directed edges
	for (int i = 0; i < edge; i++) {
		cin >> a >> b;
		// edge a -> b
		Graph[a].push_back(b);
	}
	detect_cycle(Graph, vertex);
	return 0;
}

/**
Input :
6 8
0 3
0 4
5 0
1 5
1 0
2 1
3 4
4 5
Output :
Cycle exists in the given graph

Time Complexity : O(v+e)
Space Complexity : O(v)
**/
